# ç”¨DESEQ2åšå·®å¼‚è¡¨è¾¾åŸºå› åˆ†æ

# 0. ç”¨åˆ°çš„è½¯ä»¶

| è½¯ä»¶ | ç‰ˆæœ¬ |
| --- | --- |
| fastp | 0.20.1 |
| hisat | 2.2.1 |
| samtools | 1.12 |
| stringtie | v2.1.6 |
| gffread | 0.12.7 |
| gffcompare | 0.12.6 |

# 1. RNAæµ‹åºçš„åºåˆ—è¿‡æ»¤å’Œæ¯”å¯¹

```bash
fastp -i Prefix.R1.fastq.gz -I Prefix.R2.fastq.gz -o Prefix.R1.clean.fastq.gz -O Prefix.R2.clean.fastq.gz -A -q 20 -u 10 -n 5 -w 4 -j summary/Prefix.json -h summary/Prefix.html
hisat2-build Ref.fa Rer
hisat2 --phred33 -p 36 Ref -1 Prefix.R1.clean.fastq.gz -2 Prefix.R2.clean.fastq.gz -S Prefix.sam --summary-file Prefix.txt
# -p, threads
samtools sort -@ 36 -o Prefix.sorted.bam Prefix.sam
# -@, threads, sometimes it dosen't work well.
```

---

# 2. stringtieåšè½¬å½•æœ¬ç»„è£…å’Œå®šé‡

<aside>
ğŸ’¡ å¯ä»¥ç”¨gffreadå°†gffè½¬ä¸ºgtf
`conda install -c bioconda gffread
gffread Bgy.gff -T -o bgy.gtf`

</aside>

## 2.1 å¯¹æ¯ä¸ªä¸ªä½“çš„è½¬å½•æœ¬è¿›è¡Œç»„è£…

<aside>
ğŸ’¡ å¦‚æœä¸åš2.2çš„å†…å®¹ï¼Œå°±åŠ ä¸Š-eï¼Œå¦‚æœç»§ç»­åšï¼Œå°±ä¸åŠ ã€‚æ­¤å¤–ä¸åŠ -eäº§ç”Ÿçš„ç»“æœä¸èƒ½ç”¨äºåç»­prepDE.pyçš„åˆ†æ

</aside>

```bash
stringtie -e -p 16 -G Ref.gtf -l Ref -o Prefix.gtf Prefix.sorted.bam
# -p, threads
# -G, reference annotation to use for guiding the assembly process
# -l, name prefix for output transcripts
# -o, output path/file name for the assembled transcripts GTF
# -e, this option directs StringTie to operate in expression estimation mode; this limits the processing of read alignments to estimating the coverage of the transcripts given with the -G option (hence this option requires -G).
```

## 2.2 æ£€æµ‹è½¬å½•æœ¬æ£€æµ‹çš„ç²¾ç¡®æ€§å’Œå®Œæ•´æ€§ï¼ˆé€‰åšï¼‰

<aside>
ğŸ’¡ è¿™ä¸€æ­¥ä¸»è¦æ˜¯åœ¨æ— å‚çš„æ—¶å€™åšï¼Œæˆ–è€…è¯´è¦é¢„æµ‹ä¸€äº›å¯å˜å‰ªåˆ‡åšåç»­åˆ†æï¼Œç”±äºstringtieä¸»è¦æ˜¯æ ¹æ®readsæ¯”å¯¹åˆ°çš„ä½ç½®æ¥é¢„æµ‹è½¬å½•æœ¬ï¼Œæ‰€ä»¥æœ‰å¯èƒ½åŒä¸€ä¸ªåŸºå› ä¼šäº§ç”Ÿå¤šä¸ªå¯å˜å‰ªåˆ‡ï¼Œè¿™ç§æƒ…å†µä¸‹æœ‰å¯èƒ½åè€Œå¯¹åç»­åˆ†ææœ‰å½±å“ã€‚åšåŸºå› ç»„çš„åˆ†æçš„æ—¶å€™å¯ä»¥ä¸åš2.2çš„å†…å®¹ï¼Œä¸€ä¸ªæœ‰å‚è€ƒä»·å€¼çš„[åšå®¢](https://www.jianshu.com/p/40c1c0606db6?utm_medium=timeline)å’Œ[github issue](https://github.com/gpertea/stringtie/issues/192)ã€‚

</aside>

### 2.2.1 åˆå¹¶è½¬å½•æœ¬

```bash
stringtie --merge -p 16 -G Ref.gtf -o Ref.merged.gtf -l Ref Ref.mergelist.txt
# -p, threads
# -G, reference annotation to include in the merging (GTF/GFF3)
# -o, output file name for the merged transcripts GTF
# -l, name prefix for output transcripts
```

```bash
# Ref.mergelist.txt
# Each line contains the name of one of the gtfs to merge.
SRR15464825.gtf
SRR15464826.gtf
SRR15464827.gtf
SRR15464828.gtf
SRR15464830.gtf
SRR15464831.gtf
SRR15497174.gtf
```

### 2.2.2 æ¯”è¾ƒå‚è€ƒè½¬å½•æœ¬å’Œåˆå¹¶çš„è½¬å½•æœ¬

é€šè¿‡è¿™ä¸€æ­¥æ¥æ£€éªŒè½¬å½•æœ¬ç»„è£…çš„ç²¾ç¡®åº¦å’Œå‡†ç¡®åº¦ã€‚å…·ä½“çš„æ£€éªŒç»“æœè¯´æ˜å¯ä»¥å‚è€ƒ[ç®€ä¹¦çš„è¿™ç¯‡è¯´æ˜](https://www.jianshu.com/p/08bb35f8e7c2)å’Œ[å®˜ç½‘çš„documentary](http://ccb.jhu.edu/software/stringtie/gffcompare.shtml)ã€‚

```bash
gffcompare -r Ref.gtf -G -o Ref Ref.merged.gtf
# -r, reference annotation file (GTF/GFF)
# -G, æ¯”è¾ƒè¾“å…¥çš„gtfä¸­æ‰€æœ‰çš„è½¬å½•æœ¬ï¼Œå³ä½¿å®ƒä»¬æœ‰å¯èƒ½æ˜¯å†—ä½™çš„
# -o, prefix of outputs
```

### 2.2.3 é‡æ–°ç»„è£…è½¬å½•æœ¬

å¦‚æœè¯´å‘ç°åˆå¹¶çš„è½¬å½•æœ¬ç›¸è¾ƒäºåŸæœ¬çš„å‚è€ƒè½¬å½•æœ¬ï¼ˆRef.gtfï¼‰æ–‡ä»¶æœ‰è¾ƒå¤šæ–°å‘ç°çš„è½¬å½•æœ¬ï¼Œåˆ™å¯ä»¥ç”¨åˆå¹¶çš„è½¬å½•æœ¬ä½œä¸ºreferenceé‡æ–°ç»„è£…è½¬å½•æœ¬ï¼Œå¦åˆ™ç›´æ¥ç”¨ä¹‹å‰ç»„è£…çš„ç»“æœå³å¯ã€‚

```bash
stringtie -e -p 16 -G Ref.merged.gtf -l Ref -A Prefix.tab -o Prefix.new.gtf Prefix.bam

```

---

# 3. å‡†å¤‡deseq2æ‰€éœ€çš„æ–‡ä»¶

stringtieä¸­æä¾›äº†è„šæœ¬ï¼Œå¯ä»¥ç›´æ¥ä½¿ç”¨ï¼Œé¦–å…ˆè¦æŠŠæ‰€æœ‰çš„gtfæŒ‰ç…§åå­—æ”¾åœ¨å¯¹åº”çš„æ–‡ä»¶å¤¹ã€‚

![æˆªå±2021-12-16 ä¸‹åˆ2.48.13.png](%E7%94%A8DESEQ2%E5%81%9A%E5%B7%AE%E5%BC%82%2029da4/%E6%88%AA%E5%B1%8F2021-12-16_%E4%B8%8B%E5%8D%882.48.13.png)

```bash
prepDE.py -i ./newgtfs/ -g Ref.gene.matrix -t Ref.transcript.matrix
# -i, a folder containing all sample sub-directories, or a text file with sample               ID and path to its GTF file on each line
# -g, where to output the gene count matrix
# -t, where to output the transcript count matrix (å¦‚æœæ²¡æœ‰åš2.2ï¼Œè¿™ä¸¤ä¸ªæ–‡ä»¶å…¶å®æ˜¯ä¸€æ ·çš„ï¼‰
```

---

# 3. deseq2åˆ†æ

```r
source("https://bioconductor.org/biocLite.R")
biocLite("DESeq2")
library("DESeq2")

# æ–°ç‰ˆçš„ré‡Œé¢å·²ç»ç”¨biomanageræ›¿ä»£äº†bioclite
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq2")
library("DESeq2")

countData <- as.matrix(read.csv("Ref.gene.matrix ", row.names="gene_id"))
# Ref.gene.matrixæ˜¯prepDE.pyç”Ÿæˆçš„æ–‡ä»¶

colData <- read.csv("PhenoData.txt", sep="\t", row.names=1)
# è®°å¾—æ£€æŸ¥åˆ†éš”ç¬¦

all(rownames(colData) %in% colnames(countData))
countData <- countData[, rownames(colData)]
all(rownames(colData) == colnames(countData))
# æ£€æŸ¥coldataå’Œcountdataçš„sample idæ˜¯å¦ä¸€ä¸€å¯¹åº”ï¼ˆTURE)

dds <- DESeqDataSetFromMatrix(countData = countData,colData = colData, design = ~ CHOOSE_FEATURE)
# è®°å¾—çœ‹designçš„è¡¨è¾¾å¼ï¼Œä»¥colDataé‡Œé¢çš„åˆ—çš„åå­—åšç»„åˆã€‚

dds <- DESeq(dds)
res <- results(dds)
resOrdered <- res[order(res$padj),]
resOrdered=as.data.frame(resOrdered)
write.table(resOrdered,"DEGoutput.txt",sep="\t")
 
```

---

# 4. å¯¹WGDäº§ç”Ÿçš„paralogåšè¡¨è¾¾é‡å·®å¼‚åˆ†æ

## 4.1 æå–æ¯ä¸ªåŸºå› çš„TPMå€¼

<aside>
ğŸ’¡ åšä¸åš2.2è¿™ä¸€æ­¥çš„è„šæœ¬æ˜¯ä¸ä¸€æ ·çš„ï¼Œè¿™é‡Œçš„æ˜¯ä¸åš2.2çš„ç‰ˆæœ¬ï¼Œå¦‚æœåšäº†2.2ï¼Œä¸åŒçš„transcriptå¯èƒ½å¯¹åº”åŒä¸€ä¸ªgeneidï¼Œè¿™ä¸€ç‚¹è¦æ³¨æ„ã€‚

</aside>

```bash
perl extractTPM.pl >Ref.TPM.matrix
```

- extractTPM.pl
    
    ```perl
    use strict;
    use warnings;
    use autodie;
    use 5.010;
    
    my %TPM;
    my $NumOfFile=-1;
    my @individual_names;
    my @GeneIds;
    my @gtfs=glob ("*gtf");
    foreach my $gtf (@gtfs) {
            $NumOfFile++;
            (my $individual_name=$gtf)=~s/\.gtf//;
            push @individual_names,$individual_name;
            open GTF,"<","$gtf";
            while (<GTF>) {
                next if /^#/;
                next if (/exon/);
    	        	$_=~m/gene_id "(?<GeneId>.*?)";.*TPM "(?<TPM>.*)";/;
                my $GeneId=$+{GeneId};
    						my $TPM=sprintf "%.2f",$+{TPM};
    						$TPM{$GeneId}[$NumOfFile]=$TPM;
                push @GeneIds,$GeneId;
            }
            close GTF;
    }
    
    my $individual_names=join(',',@individual_names);
    say "gene_id,$individual_names";
    my %bak;
    map {$bak{$_}++} @GeneIds;
    @GeneIds=keys %bak;
    foreach my $GeneId (@GeneIds) {
            my $exp=join(',',$GeneId,@{$TPM{$GeneId}});
            say $exp;
    }
    ```
    

## 4.2 åšWilcoxonæˆå¯¹ç§©å’Œæ£€éªŒ

### 4.2.1 æ£€éªŒ

```perl
perl wilcoxon.pair.pl [TPMMatrixFile] [KsMedianFile] [T0File] [CollinearityFile]
```

- wilcoxon.pair.pl
    
    ```perl
    ###     æœ¬è„šæœ¬ç”¨æ¥åšwilcoxonæˆå¯¹ç§©å’Œæ£€éªŒ
    ###     T0å€¼é€šè¿‡æŸ¥è¡¨æ ¹æ®æ ·æœ¬å¤§å°è·å¾—
    ###     è®°å¾—ä¿®æ”¹39è¡Œksä¸­ä½æ•°çš„åŒºé—´
    ###     è®°å¾—ä¿®æ”¹ç¬¬56è¡Œè¾“å‡ºæ–‡ä»¶çš„åå­—
    
    use 5.012;
    use strict;
    use warnings;
    use autodie;
    
    die "Usage: perl $0 [TPMMatrixFile] [KsMedianFile] [T0File] [CollinearityFile]\n" if @ARGV<4;
    
    my $TPMMatrixFile=shift;
    my $KsMedianFile=shift;
    my $T0File=shift;
    my $CollinearFile=shift;
    
    # TPM data
    open TPM,"<","$TPMMatrixFile";
    my %Exp;
    while (<TPM>) {
            next if /^gene/;
    		    s/evm\.model\.Bruguiera_gymnorrhiza_PacBio_hic_scaffold_/bgy\./g;
            s/\s+$//;
            my @eles=split/,/;
            my $GeneId=shift @eles;
            $Exp{$GeneId}=[@eles];
    }
    close TPM;
    
    # ç­›é€‰è¿‘æœŸwgdäº§ç”Ÿçš„block
    open MEDIAN,"<","$KsMedianFile";
    my @RecentWGDBlocks;
    while (<MEDIAN>) {
                    s/\s+$//;
                    my ($block,$ks)=split/\s+/;
                    $block=~m/(\d+)$/;
                    $block=$1;
                    push @RecentWGDBlocks,$block if ($ks<=0.8 && $ks>=0.3);
    }
    close MEDIAN;
    
    # T0çš„é˜ˆå€¼
    open TVALUE,"<","$T0File";
    my %TValue;
    while (<TVALUE>) {
    	s/\s+$//;
    	my ($n,$t0)=split/\s+/;
    	$TValue{$n}=$t0;
    }
    close TVALUE;
    
    # Extract WGD retention gene pairs.
    open COL,"<","$CollinearFile";
    open OUT,">","TPM.wilcoxon.txt";
    say OUT "# lt: lower than\n# gt: greater than\n#ns: no significance";
    L: while (<COL>) {
            next if /^#/;
            $_=~m/^\s?(\d+)\-/;
            my $block=$1;
            next if (($block~~@RecentWGDBlocks)==0);
            $_=~m/(bgy.*?)\s+(bgy.*?)\s+/;
            my $Gene1=$1;
            my $Gene2=$2;
            next L if (!exists $Exp{$Gene1});
            next L if (!exists $Exp{$Gene2});
            WilcoxonPairedTest ($Gene1,$Gene2);
    }
    close COL;
    close OUT;
    
    sub WilcoxonPairedTest {
            my $Gene1=shift;
            my $Gene2=shift;
            my @Gene1Exp=@{$Exp{$Gene1}};
            my @Gene2Exp=@{$Exp{$Gene2}};
            my $Gene1Exp=join(',',@{$Exp{$Gene1}});
            my $Gene2Exp=join(',',@{$Exp{$Gene2}});        
    print "$Gene1\n$Exp{$Gene1}[0]\n$Gene1Exp[0]\n$Gene2\n$Exp{$Gene2}[0]\n$Gene2Exp[0]\n";
            my (@Differences,@ABSDifferences);
            my ($Rank,@RankOrder);
    		my $PositiveRankSum=0;
    		my $NegativeRankSum=0;
    # æ±‚è¡¨è¾¾é‡çš„å·®å€¼
            for (1..@Gene1Exp) {
                    my $Gene1Exp=shift @Gene1Exp;
                    my $Gene2Exp=shift @Gene2Exp;
                    my $Difference=$Gene1Exp-$Gene2Exp;
                    push @Differences,$Difference;
            }
    # æ’åºå¹¶è®°å½•ç»å¯¹å€¼
            @Differences=sort {abs($a)<=>abs($b)} @Differences;
            @ABSDifferences=map {abs ($_)} @Differences;
            my %num;
            map {$num{$_}++} @ABSDifferences;
    # å·®å€¼ä¸º0çš„è¯åœ¨æ’åºçš„æ—¶å€™è¦åˆ é™¤
            delete $num{'0'};
    		my @NonzeroValues=keys %num;
    		my $NonzeroValueNum=@NonzeroValues;
    		next L if $NonzeroValueNum<6;
            $Rank=0;
            @RankOrder;
            @ABSDifferences=keys %num;
    # æ ¹æ®å·®å€¼çš„ç»å¯¹å€¼æ±‚ç§©
            foreach my $ABSDifference (@ABSDifferences) {
                    if ($num{$ABSDifference}==1) {
                            $Rank+=1;
                            push @RankOrder,$Rank;
                    } else {
                            my $RankLow=$Rank+1;
                            my $RankHigh=$Rank+$num{$ABSDifference};
                            my $AverageRank=($RankLow+$RankHigh)/2;
                            for (1..$num{$ABSDifference}) {push @RankOrder,$AverageRank};
                            $Rank=$RankHigh;
                    }
            }
            foreach my $Difference (@Differences) {
                    next if ($Difference==0);
                    my $Rank=shift @RankOrder;
                    $PositiveRankSum+=$Rank if $Difference>0;
                    $NegativeRankSum+=$Rank if $Difference<0;
            }
            my $LowerRankSum=($PositiveRankSum>$NegativeRankSum)?($NegativeRankSum):($PositiveRankSum);
            if ($LowerRankSum<=$TValue{$NonzeroValueNum}) {
                    my $RankSumStat=($PositiveRankSum>$NegativeRankSum)?('gt'):('lt');
                    say OUT "$Gene1\t$Gene1Exp\t$Gene2\t$Gene2Exp\t$RankSumStat";
            } else {
                    say OUT "$Gene1\t$Gene1Exp\t$Gene2\t$Gene2Exp\tns";
            }
    }
    ```
    

### 4.2.2 ç”»å›¾

<aside>
ğŸ’¡ å…ˆæŠŠä¸Šä¸€æ­¥è¾“å‡ºçš„ç»“æœæ•´ç†æˆ`TPM.matrix.forheatmap.txt` å½¢å¼ï¼Œç„¶åæœ‰ä¸€ä¸ªæ–‡ä»¶æŒ‡å®šåˆ—çš„åˆ†ç»„ï¼Œå¦‚SampleGroup.txtï¼Œä¹‹åç”¨TBtoolsç”»å›¾å³å¯ï¼Œæ³¨æ„å¯ä»¥åˆ†ä¸€ä¸ªmaingroupå’Œä¸€ä¸ªsubgroupã€‚

</aside>

- TPM.matrix.forheatmap.txt
    
    ![Untitled](%E7%94%A8DESEQ2%E5%81%9A%E5%B7%AE%E5%BC%82%2029da4/Untitled.png)
    
- SampleGroup.txt
    
    ![Untitled](%E7%94%A8DESEQ2%E5%81%9A%E5%B7%AE%E5%BC%82%2029da4/Untitled%201.png)
    

## 4.3 æ•´ç†wilcoxonç§©å’Œæ£€éªŒå’Œå·®å¼‚è¡¨è¾¾åŸºå› çš„ç»“æœ

<aside>
ğŸ’¡ æ³¨æ„æ–‡ä»¶ä¸­åŸºå› åè¦ä¸€ä¸€å¯¹åº”ï¼Œæˆ–è€…æ›´æ”¹ç¬¬ä¸€ä¸ªå¾ªç¯ä¸­æ³¨é‡Šæ‰çš„é‚£è¡Œä¸­çš„æ­£åˆ™ã€‚

</aside>

### 4.3.1 çœ‹ç§©å’Œæ£€éªŒä¸­æœ‰è¡¨è¾¾å·®å¼‚çš„åŸºå› å¯¹åœ¨ç›èƒè¿«ä¸‹æ˜¯å¦æœ‰å·®å¼‚è¡¨è¾¾

```bash
perl wilcoxon_deg.pl DEGoutput.txt TPM.matrix.forheatmap.txt
```

- wilcoxon_deg.pl
    
    ```perl
    open DEG,"<","$ARGV[0]";
    open WGD,"<","$ARGV[1]";
    my %ExpDifference;
    
    while (<DEG>) {
        next if /baseMean/;
    # s/evm\.model\.Bruguiera_gymnorrhiza_PacBio_hic_scaffold_/bgy\./;
        my ($genename,$log2fc)=(split/\s+/)[0,2];
    		$genename=~s/"//g;
        $ExpDifference{$genename}='Up' if $log2fc<-1;
        $ExpDifference{$genename}='Down' if $log2fc>1;
    }
    close DEG;
    
    open OUT,">DEG.wilcoxon.txt";
    
    while (<WGD>) {
        next if /GeneID/;
        my $genepair=(split/\s+/)[0];
        my ($gene1,$gene2)=split/_/,$genepair;
        print OUT  "$gene1\t$gene2\t";
        map {
            if (exists $ExpDifference{$_}) {print OUT "$ExpDifference{$_}\t"} else {print OUT "NS\t"};
        } ($gene1,$gene2);
            print OUT "\n";
    }
    close WGD;
    close OUT;
    ```
    

### 4.3.2 å»é™¤å‡ä¸å·®å¼‚è¡¨è¾¾çš„åŸºå› å¯¹

```bash
perl remove.doubleNS.pl DEG.wilcoxon.txt >DEG.wilcoxon.filtered.txt
```

- remove.doubleNS.pl
    
    ```perl
    while (<>) {
    	my $line=$_;
    	s/\s+$//;
    	my ($sig1,$sig2)=(split/\s+/)[2,3];
    	next if ($sig1 eq 'NS' && $sig2 eq 'NS');
    	print $line
    }
    ```